scripts/mod_loader/ui/widgets/draggable.lua

```diff
@@ -76,8 +76,20 @@ function UiDraggable:startDrag(mx, my, button)
 		self.__resizeDir = getResizeDirection(self, mx, my, self.__resizeHandle)
 	end
 
-	if self.parent then
-		self.parent:relayout()
+	if self.dragPlaceholder ~= nil then
+		local root = sdlext.getUiRoot()
+		local owner = self.parent
+		local index = list_indexof(owner.children, self)
+		self:detach()
+		self.owner = owner
+		self.translucent = true
+
+		owner:add(self.dragPlaceholder, index)
+		self.dragPlaceholder:show()
+
+		self:addTo(root.draggableUi)
+		self.x = self.screenx
+		self.y = self.screeny
```

Comment:
I'm not entirely sold on the idea of placeholder management being a responsibility of the baseline `UiDraggable`... To me, the notion of a placeholder is an implementation detail of only one usecase of the drag'n'drop system - one that exists only to account for a quirk of layouts.  
It feels like this should be handled by a separate helper function/class that implements placeholders on top of a generic drag'n'drop framework.

Reply:
I can't see I completely see your issue, but I suppose I can see it potentially becoming one, if you envision drag-dropping being used in many other ways down the line.

So for a separate class, I can have this suggested solution. Would this be fine? (tested, and should be identical to the code in the PR except pulled out to a separate class)
```lua
UiDragDropList = Class.inherit(UiDraggable)

function UiDragDropList:new()
	UiDraggable.new(self)
end

function UiDragDropList:startDrag(mx, my, button)
	if button ~= 1 then
		return
	end

	UiDraggable.startDrag(self, mx, my, button)

	if self.dragPlaceholder ~= nil then
		local root = sdlext.getUiRoot()
		local owner = self.parent
		local index = list_indexof(owner.children, self)
		self:detach()
		self.owner = owner
		self.translucent = true

		owner:add(self.dragPlaceholder, index)
		self.dragPlaceholder:show()

		self:addTo(root.draggableUi)
		self.x = self.screenx
		self.y = self.screeny
	end
end

function UiDragDropList:stopDrag(mx, my, button)
	if button ~= 1 then
		return
	end

	if self.dragMoving and self.dragPlaceholder ~= nil then
		local index = list_indexof(self.owner.children, self.dragPlaceholder)
		self.dragPlaceholder:detach()
		self.dragPlaceholder:hide()

		self:detach()
		self:addTo(self.owner, index)
		self.translucent = self.dragPlaceholder.translucent
	end

	UiDraggable.stopDrag(self, mx, my, button)
end

function Ui:registerDragDropList(placeholder)
	Assert.NotEquals('nil', type(placeholder), "Argument #1: Must be an ui element")
	self:registerDragMove()
	self.dragPlaceholder = placeholder
	self.startDrag = UiDragDropList.startDrag
	self.stopDrag = UiDragDropList.stopDrag
end
```

Reply:
A sidenote: Maybe there should be an `Assert.IsClass` function?

Reply:
Yeah, that's class is pretty much what I had in mind.

Apart from a vague notion of what would make sense responsibility-wise, my main concern was with `stopDrag` reinserting the placeholder into the owner element, since with the `addTo` call, we depend on the target element's layout logic to correctly translate the element.

This works for elements that manage their children's positions (most/all of the `___Layout` classes), but breaks for elements that are just the bare `Ui` class (which is the reason we need to translate the placeholder when moving it to `root.draggableUi`). `UiDraggable` simply has insufficient information to account for that, since it doesn't really know how the parent element is going to behave, and when a correction is and isn't needed.

Regarding the assertions, I'm not completely against them, but we can already kinda do that via `Assert.True(Class.instanceOf(placeholder, Ui))`, and other cases via `isSuperclassOf`/`isSubclassOf`. I don't think having dedicated assertions for these would net much benefit, apart from being slightly shorter. Though these functions *could* use updating to not use base lua `assert`, to make them more informative in failure cases.

Reply:
I see. With that explanation I can certainly agree that the placeholder add/detach logic would not be the right fit for every type of parent element. Good call on moving it to another class.

Regarding the Asserts I was suggesting, I was not concerned with the longer code by using `Assert.True`. The reason I suggested `Assert.IsClass` (or a variation of it) was mostly to provide clearer error messages than `"[msg]: Expected true, but was false"`, which in itself doesn't say much of anything.

On further thinking though, classes don't know their own name, so it would in any case have been provided to the `Assert.IsClass` function. Using `Assert.True` I suppose maybe `"Argument#1 - IsSubClassOf(Ui): Expected true, but was false"` would be a sufficient solution for class checks.